pulsar-io/nifi/src/main/java/org/apache/pulsar/io/nifi/NiFiSink.java

```diff
@@ -0,0 +1,106 @@
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pulsar.io.nifi;
+
+import lombok.extern.slf4j.Slf4j;
+import org.apache.nifi.remote.Transaction;
+import org.apache.nifi.remote.TransferDirection;
+import org.apache.nifi.remote.client.SiteToSiteClient;
+import org.apache.nifi.remote.client.SiteToSiteClientConfig;
+import org.apache.pulsar.functions.api.Record;
+import org.apache.pulsar.io.core.Sink;
+import org.apache.pulsar.io.core.SinkContext;
+import org.apache.pulsar.io.core.annotations.Connector;
+import org.apache.pulsar.io.core.annotations.IOType;
+
+import java.io.IOException;
+import java.util.Map;
+
+
+/**
+ * A sink that delivers data to Apache NiFi using the NiFi Site-to-Site client. The sink requires
+ * a NiFiDataPacketBuilder which can create instances of NiFiDataPacket from the incoming data.
+ */
+@Connector(
+    name = "nifi",
+    type = IOType.SINK,
+    help = "The NiFiSink is used for moving messages from Pulsar to Apache NiFi using the NiFi Site-to-Site client.",
+    configClass = NiFiConfig.class
+)
+@Slf4j
+public class NiFiSink implements Sink<NiFiDataPacket> {
+
+    private NiFiConfig niFiConfig;
+    private SiteToSiteClient client;
+    private SiteToSiteClientConfig clientConfig;
+
+    @Override
+    public void open(Map<String, Object> config, SinkContext sinkContext) throws Exception {
+        niFiConfig = NiFiConfig.load(config);
+        if (niFiConfig.getUrl() == null
+                || niFiConfig.getPortName() == null
+                || niFiConfig.getRequestBatchCount() <=0) {
+            throw new IllegalArgumentException("Required property not set.");
+        }
+
+        clientConfig = new SiteToSiteClient.Builder()
+                .url(niFiConfig.getUrl())
+                .portName(niFiConfig.getPortName())
+                .requestBatchCount(niFiConfig.getRequestBatchCount())
+                .buildConfig();
+        client = new SiteToSiteClient.Builder().fromConfig(clientConfig).build();
+    }
+
+    @Override
+    public void write(Record<NiFiDataPacket> record) {
+        Transaction transaction = null;
+        try {
+            transaction = client.createTransaction(TransferDirection.SEND);
+        } catch (IOException e) {
+            log.error("create Transaction exception ", e);
+        }
+        if (transaction == null) {
+            throw new IllegalStateException("Unable to create a NiFi Transaction to send data");
+        }
+        NiFiDataPacket niFiDataPacket = record.getValue();
+
+        if (log.isDebugEnabled()) {
+            log.debug("Record sending to kafka, record={}.", record);
+        }
+
+        try {
+            transaction.send(niFiDataPacket.getContent(), niFiDataPacket.getAttributes());
+            record.ack();
+
+            transaction.confirm();
+            transaction.complete();
```

Comment:
> I have a performance concern here. `write` method ideally should be non-blocking. otherwise it would hurt throughput. if I understand correctly here, `complete` would be blocking call. can we try to batch records into one transaction and complete the transaction in a background thread?
 
I try this comment code in new commit.